registry/config.go

```diff
@@ -63,6 +65,10 @@ var (
 	emptyServiceConfig = newServiceConfig(ServiceOptions{})
 )
 
+var (
+	validHostnameRegex = regexp.MustCompile(`^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])$`)
```

Comment:
We should allow `_` characters. While the DNS RFC says they should not be used in hostnames, they are often used in practice, especially outside the context of DNS (i.e. `/etc/hosts`).